include/deal.II/matrix_free/operators.h

```diff
@@ -205,6 +361,260 @@ namespace MatrixFreeOperators
       }
   }
 
+  //----------------- Base operator -----------------------------
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  Base<dim,fe_degree,n_q_points_1d,number>::~Base ()
+  {
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  Base<dim,fe_degree,n_q_points_1d,number>::Base ()
+    :
+    Subscriptor(),
+    data(NULL),
+    have_interface_matrices(false)
+  {
+  }
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  unsigned int
+  Base<dim,fe_degree,n_q_points_1d,number>::m () const
+  {
+    Assert(data != NULL,
+           ExcNotInitialized());
+    return data->get_vector_partitioner()->size();
+  }
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  unsigned int
+  Base<dim,fe_degree,n_q_points_1d,number>::n () const
+  {
+    return m();
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  void
+  Base<dim,fe_degree,n_q_points_1d,number>::clear ()
+  {
+    inverse_diagonal_entries.reinit(0);
+    diagonal_entries.reinit(0);
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  number
+  Base<dim,fe_degree,n_q_points_1d,number>::el (const unsigned int row,
+                                                const unsigned int col) const
+  {
+    Assert (row == col, ExcNotImplemented());
+    Assert (diagonal_entries.size() > 0, ExcNotInitialized());
+    return diagonal_entries(row);
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  void
+  Base<dim,fe_degree,n_q_points_1d,number>::initialize_dof_vector (LinearAlgebra::distributed::Vector<number> &vec) const
+  {
+    Assert(data != NULL,
+           ExcNotInitialized());
+    if (!vec.partitioners_are_compatible(*data->get_dof_info(0).vector_partitioner))
+      data->initialize_dof_vector(vec);
+    Assert(vec.partitioners_are_globally_compatible(*data->get_dof_info(0).vector_partitioner),
+           ExcInternalError());
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  void
+  Base<dim,fe_degree,n_q_points_1d,number>::
+  initialize (const MatrixFree<dim,number> &data_)
+  {
+    data = &data_;
+    edge_constrained_indices.clear();
+    have_interface_matrices = false;
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename number>
+  void
+  Base<dim,fe_degree,n_q_points_1d,number>::
+  initialize (const MatrixFree<dim,number> &data_,
+              const MGConstrainedDoFs &mg_constrained_dofs,
+              const MPI_Comm &mpi_communicator,
```

Comment:
I think it would be better to remove `mpi_communicator` from this call (it is a bit strange to have a communicator here but not for the other method). The right thing to do would be to query the MPI communicator from the vector partitioner, `data_.get_vector_partitioner()->get_communicator()`, or alternatively, from the triangulation - but in that case you need to downcast to a parallel triangulation.